src/importlinter/contracts/_common.py

```diff
@@ -18,7 +18,8 @@
 class Link(TypedDict):
     importer: str
     imported: str
-    line_numbers: Tuple[int, ...]
+    # If the graph has been built manually, we may not know the line number.
+    line_numbers: Tuple[Optional[int], ...]
```

Comment:
The problem with doing this is that now all code must check every single number in the sequence of line numbers to see if it's not null, before it can use them :(

Reply:
But that's true, isn't it? We're just making it more obvious.

`graph.get_import_details` is [not guaranteed to return any details](https://grimp.readthedocs.io/en/stable/usage.html#ImportGraph.get_import_details) even if there is an import there. While in practice this is not the case within Import Linter at the moment, it feels to me sensible to treat the graph we're analysing as if that is a possibility. `(None,)` is IMO a better way of encoding missing import details that `(-1,)`.

The other aspect of this is that at the moment Grimp doesn't explicitly tell us if import details are missing, but in the future this might change.

Reply:
I guess it makes sense if some line numbers may be missing while others are not missing, but if it's always one or the other, it would be better to just mark the `line_numbers` property in full as nullable. That way you only have to check one thing, not many things.

Reply:
At the moment it's always one or the other, but I think this is a limitation of Grimp that could change in the future.